### PR TITLE
mingw64 port: networking, other fixes

### PR DESCRIPTION
**Pull request by eudoxos**
Sorry for closing the previous PR, I made some git mess rebasing, so here we go again, this time with code that actually works. With wine, I can enumerate network addresses with arv-network-test.exe

```
           interface      proto          address             mask        broadcast
                  lo       IPv4        127.0.0.1        255.0.0.0        127.0.0.1
                eth3       IPv4     169.254.64.1      255.255.0.0     169.254.64.1
                eth3       IPv4    192.168.100.1      255.255.0.0    192.168.100.1
                eth3       IPv4     192.168.2.12      255.255.0.0     192.168.2.12
              virbr0       IPv4    192.168.122.1    255.255.255.0    192.168.122.1
             docker0       IPv4       172.17.0.1      255.255.0.0       172.17.0.1
     br-23a99ce39657       IPv4       172.18.0.1      255.255.0.0       172.18.0.1
```

For the record, but this is not my ambition yet, I see this with arv-tool-0.8.exe:
```
00f0:err:winsock:interface_bind Failed to bind to interface, receiving broadcast packets will not work on socket 009c.
00f0:err:winsock:interface_bind Failed to bind to interface, receiving broadcast packets will not work on socket 00a8.
No device found
```

This should not be too difficult to fix later.

### CODE REVIEW

**Comment by eudoxos** [comment]
It actually runs in virtual Windows 10 (those funny characters, no idea, ...). arv-tool does not talk to the camera yet, though the address can be pinged. I suspect it will be related to the broadcast issue which wine reported (but windows perhaps don't). I will test on a real machine later.

image

File: src/arvnetwork.c
Comment (EmmanuelP) [inline src/arvnetwork.c:51]: Please define a destructor and call WSACleanup from there

File: src/arvnetwork.c
Comment (EmmanuelP) [inline src/arvnetwork.c:72]: If WSAStartup is mandatory, use g_critical instead of fprintf

File: src/arvnetwork.c
Comment (EmmanuelP) [inline src/arvnetwork.c:43]: Is it really mandatory ? I've never added declaration when I've used constructors.

**Review by EmmanuelP — CHANGES_REQUESTED** [review]

**Comment by eudoxos** [comment]
Address reporting is fixed, as well as building with all features (provided all dependencies - which are all packaged by MSYS2 already) are installed: documentation, usb, gst-plugin, viewer. There is still an error with introspection, where the linker fails to find `-laravis-0.8`).

File: src/arvnetwork.c
Comment (eudoxos) [reply src/arvnetwork.c:43]: Yes, it is (also stated in [WSAStartup docs](https://docs.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-wsastartup)) . It was the reason for the funny characters as `getnameinfo` was failing (saying the `WSAStartup` was not called yet). This only manifested in real Windows 10 installation, not in wine.

Comment (eudoxos) [reply src/arvnetwork.c:43]: I get this warning without the declaration:
```
../src/arvnetwork.c:47:6: warning: no previous prototype for ‘arv_initialize_networking’ [-Wmissing-prototypes]
   47 | void arv_initialize_networking(void){
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~
../src/arvnetwork.c:80:6: warning: no previous prototype for ‘arv_cleanup_networking’ [-Wmissing-prototypes]
   80 | void arv_cleanup_networking(void){
      |      ^~~~~~~~~~~~~~~~~~~~~~
```

**Comment by eudoxos** [comment]
The cause of failing introspection is that meson calls guest's `g-ir-inspect`. Fixing this might require proper meson cross-compilation setup (perhaps not complicated, just never tried). I am currently using [crossroad](https://pypi.org/project/crossroad/) which defines wrapper binaries instead (and does not provide support for `g-ir-inspect`).

File: src/arvnetwork.c
Comment (EmmanuelP) [reply src/arvnetwork.c:43]: I've just remembered constructor and destructor should be made static.

**Comment by eudoxos** [comment]
I am having issue with receiving broadcast traffic. I have a camera (`169.194.21.103`) which is responding to the broadcast sent using the host (Linux) arv-tool (through `169.254.64.1`):

image

When I run arv-tool from Windows 10 (with bridged networking, hence the same interface are available) or from wine on Linux, I see this:
image

I was checking in `arvgvinterface.c` in `_discover`, no packet is ever seen by aravis, it just times out waiting for the reply.

The output I get in wine is (NB: `Failed to bind` is reported for 192.168.100.1 and 192.168.2.12, not for the 169.254.64.1):
```
ARV_DEBUG=all wine arv-tool-0.8.exe 
ERROR: ld.so: object 'libgtk3-nocsd.so.0' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
ERROR: ld.so: object 'libgtk3-nocsd.so.0' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
WSAStartup done.
[GvDiscoverSocket::new] Add interface 127.0.0.1
[GvDiscoverSocket::new] Add interface 169.254.64.1
[GvDiscoverSocket::new] Add interface 192.168.100.1
015c:err:winsock:interface_bind Failed to bind to interface, receiving broadcast packets will not work on socket 009c.
[GvDiscoverSocket::new] Add interface 192.168.2.12
015c:err:winsock:interface_bind Failed to bind to interface, receiving broadcast packets will not work on socket 00a8.
[GvDiscoverSocket::new] Add interface 192.168.122.1
[GvDiscoverSocket::new] Add interface 172.17.0.1
[GvDiscoverSocket::new] Add interface 172.18.0.1
```

Any hint on where to look?

File: src/arvnetwork.c
Comment (eudoxos) [reply src/arvnetwork.c]: Fixed.

**Comment by EmmanuelP** [comment]
> Any hint on where to look?

May be g_poll fails because of the socket in errors ?

I have just a look at _discover, and found 2 issues:

* error code is not handled and leaked if it is set to a non NULL value by g_socket_bind.
 * in case of error, the corresponding socket should not be added to the list.

**Comment by EmmanuelP** [comment]
And in _discover, the case where g_poll returns -1 (error) is not handled.

**Comment by eudoxos** [comment]
> And in _discover, the case where g_poll returns -1 (error) is not handled.

I was looking at that already, it times out properly, without error.

I tried again without virtual interface (ine IP only), it times out just the same.

**Comment by eudoxos** [comment]
I think I am on the track now, [g_socket documentation](https://developer.gnome.org/gio/stable/GSocket.html#g-socket-bind) says:

> On POSIX systems, the file descriptors in fds can be any sort of file descriptor, but the situation is much more complicated on Windows. If you need to use g_poll() in code that has to run on Windows, the easiest solution is to construct all of your GPollFDs with g_io_channel_win32_make_pollfd().

**Comment by eudoxos** [comment]
I am opening a separate repository with minified broadcast-discovery (exhibiting the same issue): https://github.com/eudoxos/bcast .

**Comment by EmmanuelP** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ src/arvgvinterface.c  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/EmmanuelP/aravis/pullRequest?prid=6356687&bid=20636766)

**Comment by eudoxos** [comment]
I have some success here, with (seemingly) straight way forwards. Let me recap first:

1. `g_poll` does not work on socket FDs under windows; socket FDs are windows handles, not regular descriptors; as `g_poll` internally calls calls [WaitForMultipleObjectsEx](https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitformultipleobjectsex) designed for regular FDs only, it fails to detect incoming packet.

2. `WaitForMultpleObjectsEx` (and `g_poll`, by extension) do work on `WSAEVENT` descriptors (which are regular FDs). `WSAEVENT` is created via [WSACreateEvent](https://docs.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-wsacreateevent) (and later closed via [WSACloseEvent](https://docs.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-wsacloseevent)), and must be attached to the socket via [WSAEventSelect](https://docs.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-wsaeventselect), specifying mask of events to watch. 

3. When `g_poll` signals data available, [WSAEnumNetworkEvents](https://docs.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-wsaenumnetworkevents) must be called so that winsock internal record of new network events is cleared (otherwise `g_poll` will signal data available over and over). This call must be done before `g_socket_receive` (or similar) so that we don't miss any data.

4. There is an ancient bug [`g_poll` does not work on win32 sockets](https://gitlab.gnome.org/GNOME/glib/-/issues/214) in gnome bugzilla with an (unaccepted) patch attached. The recommended solution is to use `GSource` and main loop (which Aravis does not use, as far as I see), which will work with win32 sockets without issues.

I currently modified those places which are used for device discovery, thus I can successfully run arv-tool under wine.

It still times out under (virtualized) Windows 10.

**Comment by eudoxos** [comment]
I can confirm now the enumeration works also under Windows 10 (I had an old aravis DLL installed somewhere which was used instead of the updated one).

**Comment by EmmanuelP** [comment]
> I can confirm now the enumeration works also under Windows 10 (I had an old aravis DLL installed somewhere which was used instead of the updated one).

Great. I will review your pull request before the end of this week.

**Comment by eudoxos** [comment]
The last bit (working around `g_poll` limitations) is a bit messy, some suggestion how to make it nicer would be welcome (besides fixing it in GLib proper): there are 3 places to modify around each use of `g_poll`. There is a [summary I wrote](https://discourse.gnome.org/t/g-poll-times-out-with-windows/4594/9) at Gnome Discourse which explains why.

**Comment by EmmanuelP** [comment]
Hi,

I'm not comfortable with the number of #ifdefs around the polling code. Also the same code seems to be repeated several times. It looks like there is a pattern emerging:

- Creation of the polling structure
- Polling
- Cleanup

Did you try to put WSCreateEvent, WSAEventSelect, WSAEnumeNetworkEvent and WSACloseEvent around the g_poll function ?

If that does not work, I guess that could be handled by a dedicated object. I'm not sure about the right naming, may be ArvPollingInfo, with the following functions:

- arv_polling_info_new
- arv_polling_info_add_fd
- arv_polling_info_wait_for_event
- arv_polling_info_free

What is the status of your work. Is everything working on Windows ?

Cheers.

**Comment by eudoxos** [comment]
Neither I am happy with the ifdefs. I [described the pattern here](https://discourse.gnome.org/t/g-poll-times-out-with-windows/4594/9). How about something like the following (untested code)? It would go to arvnetwork.c, non-Windows versions would be no-op.


```c
/* modify GPollFD array before g_poll is called */
void
arv_gpollfd_prepare_all(GPollFD *fds, guint nfds){
	guint n;
	int wsaRes;
	for (n = 0; n < nfds; n++){
		gint64 fd = fds[i].fd;
		fds[i].fd = (gint64) WSACreateEvent();
		g_assert (fds[i].fd != WSA_INVALID_EVENT);
		wsaRes = WSAEventSelect (fd, fds[i].fd, FD_READ);
		g_assert (wsaRes == 0);
	}
}

/*
clear flag on the socket after g_poll returns;
unfortunately we cannot recover societ fd from GPollFD, it must be passed as second arg.
Perhaps return gboolean to indicate an error condition, or use GError* (?)
*/
void
arv_gpollfd_clear_one(GPollFD *fd, GSocket* socket){
	WSANETWORKEVENTS wsaNetEvents;
	int wsaRes;
	wsaRes = WSAEnumNetworkEvents (g_socket_get_fd(socket), fds.fd, &wsaNetEvents);
	/* TODO: check return value, check wsaNetEvents for errors */
}

/* release event descriptors for the entire array of GPollFDs */
void
arv_gpollfd_finish_all(GPollFD *fds, guint nfds){
	for (i = 0; i < nfds; i++){
		WSACloseEvent(fds[i].fd);
	}
}

```

The Windows version still does not receive stream packets, cause of that is unknown to me (`g_poll` times out); but I hope it is only a minor detail once this is settled.

**Comment by EmmanuelP** [comment]
Thanks for the last commit, that is way better now. Please document add a comment before the functions on why they exist, with pointer to the bug report and the discourse topic.

**Comment by eudoxos** [comment]
> Thanks for the last commit, that is way better now. Please document add a comment before the functions on why they exist, with pointer to the bug report and the discourse topic.

Good :) I will test more tomorrow, write some docs and commit the polished version.

**Comment by EmmanuelP** [comment]
Hi,

> The Windows version still does not receive stream packets, cause of that is unknown to me (g_poll times out);

I don't get it. what is the status of the stream reception ?

**Comment by eudoxos** [comment]
> I don't get it. what is the status of the stream reception ?
`g_poll` times out (incomplete packet received with `ARV_DEBUG` or somesuch — can't check now) for stream reception (other polls, like in discovery, work).

**Comment by eudoxos** [comment]
Hi Emmanuel, would you consider merging this PR in the current half-functional state? I won't have time to work on it now, it seems, but maybe someone else will feel motivated, or I will later. There are some cleanups which als non-windows code profits from. It will also prevent divergence between branches, if they are merged sooner than later.

**Comment by EmmanuelP** [comment]
Hi,

First of all, thanks for your interrest in Aravis and your work on the Windows support. I hope you will be able to resume it.

I was not able yet to try the compilation for Windows, and I don't exactly know the status of this patch serie.

Is it possible to capture image buffers from a GigEVision camera ?
Is it possible to capture image buffers from a USB3Vision camera ?
Is it possible to list the available cameras ?

**Comment by eudoxos** [comment]
> First of all, thanks for your interrest in Aravis and your work on the Windows support. I hope you will be able to resume it.

You are the first one who should be praised for this project, really. Thanks :)

> Is it possible to list the available cameras ?

This works fine (thanks to the networking fixes and `arv_gpollfd_*` calls around `g_poll`).

> Is it possible to capture image buffers from a GigEVision camera ?

Stream reception fails (`g_poll` never signals packet arriving, even though `arv_gpollfd_*` are used just like elsewhere).

> Is it possible to capture image buffers from a USB3Vision camera ?

Did not try (yet), might do with an a Basler USB3 camera. #431 is another attempt at that.

**Comment by eudoxos** [comment]
This is a note for @stu85010 (and others) who might want to join forces (https://github.com/stu85010/glib_w32_udp_listener/issues/2).

The current situation is puzzling in that discovery works (meaning we are generally able to receive UDP traffic on a socket via `g_poll`, with workarounds needed for win32 as described above) whereas stream reception is dysfunctional — the packets do arrive (seen in Wireshark), the port is open, the `g_poll`'d socket is prepared the same way as the discovery socket (not 100% exactly though, it was not used for broadcasting the discovery packet prior to listening) but `g_poll` here is deaf.

**Comment by eudoxos** [comment]
I am slowly moving forward (hopefully). Now the packet size is negotiated (tried) and I am tracing how are the fragments processed.

I am testing with arv-viewer, and I am seeing this after I start the stream (with `ARV_DEBUG=stream:3,device:3`). 

What does that tell you, with your expert eye?



```
  bins  ;Buffer r
       0;       0
    2000;       0
    4000;       0
    6000;       0
    8000;       0
   10000;       0
   12000;       0
   14000;       0
   16000;       0
   18000;       0
   20000;       0
   22000;       0
   24000;       0
   26000;       0
   28000;       0
   30000;       0
   32000;       0
   34000;       0
   36000;       0
   38000;       0
   40000;       0
   42000;       0
   44000;       0
   46000;       0
   48000;       0
   50000;       0
   52000;       0
   54000;       0
   56000;       0
   58000;       0
   60000;       0
   62000;       0
   64000;       0
   66000;       0
   68000;       0
   70000;       0
   72000;       0
   74000;       0
   76000;       0
   78000;       0
   80000;       0
   82000;       0
   84000;       0
   86000;       0
   88000;       0
   90000;       0
   92000;       0
   94000;       0
   96000;       0
   98000;       0
  100000;       0
  102000;       0
  104000;       0
  106000;       0
  108000;       0
  110000;       0
  112000;       0
  114000;       0
  116000;       0
  118000;       0
  120000;       0
  122000;       0
  124000;       0
  126000;       0
  128000;       0
  130000;       0
  132000;       0
  134000;       0
  136000;       0
  138000;       0
  140000;       0
  142000;       0
  144000;       0
  146000;       0
  148000;       0
  150000;       0
  152000;       0
  154000;       0
  156000;       0
  158000;       0
  160000;       0
  162000;       0
  164000;       0
  166000;       0
  168000;       0
  170000;       0
  172000;       0
  174000;       0
  176000;       0
  178000;       0
  180000;       0
  182000;       0
  184000;       0
  186000;       0
  188000;       1
-------------
>=190000;       8
<      0;       0
min     ;  189269
max     ; 2192775
last max
at:     ;       2
Counter =       16
[GvStream::finalize] n_completed_buffers    = 1
[GvStream::finalize] n_failures             = 14
[GvStream::finalize] n_timeouts             = 14
[GvStream::finalize] n_aborteds             = 0
[GvStream::finalize] n_underruns            = 16930
[GvStream::finalize] n_missing_frames       = 5
[GvStream::finalize] n_size_mismatch_errors = 0
[GvStream::finalize] n_received_packets     = 40640
[GvStream::finalize] n_missing_packets      = 19259
[GvStream::finalize] n_error_packets        = 0
[GvStream::finalize] n_ignored_packets      = 28517
[GvStream::finalize] n_resend_requests      = 374791
[GvStream::finalize] n_resent_packets       = 2863
[GvStream::finalize] n_resend_ratio_reached = 5
[GvStream::finalize] n_duplicated_packets   = 2997
[Stream::finalize] Flush 5 buffer[s] in input queue
[Stream::finalize] Flush 0 buffer[s] in output queue
```

**Comment by EmmanuelP** [comment]
Hi,

There is a lot of resend requests, 15 times the number of missing packets, which is a lot. The packet resend mechanism relies on the time returned by the system. May be the time resolution is not small enough on Windows ?

Also, you should test aravis using arv-camera-test. The viewer user interface may mess with the stream thread.

You may want to set a 1 Hz frame rate, with a small image size, 128x128 for example, and increase de debug level to 4.

**Comment by eudoxos** [comment]
Good idea, arv-camera-test is more convenient (no clicking). This is the output (the timeout debug messages after `g_poll`s are added by myself locally). I will try with the resolution changed later.
```
$ ./arv-camera-test.exe
Looking for the first available camera
002b:err:winsock:interface_bind Failed to bind to interface, receiving broadcast packets will not work on socket 009c.
[GvDevice::new] Interface address = 169.254.21.100
[GvDevice::new] Device address = 169.254.21.101
[GvDevice::load_genicam] xml url = 'Local:STEMMERIMAGING_CVGevServer_rev2_775_0.xml;C3B0;47A50' at 0x200
[GvDevice::load_genicam] Xml address = 0xc3b0 - size = 0x47a50 - STEMMERIMAGING_CVGevServer_rev2_775_0.xml
[GvDevice::new] Packet resend = yes
[GvDevice::new] Write memory = yes
[GvDevice::new] Legacy endianness handling = yes
vendor name           = VRmagicImaging
model name            = VRmLS1-GEV#128#018-660#C
device id             = EFD0AFMH07
image width           = 2048
image height          = 1088
horizontal binning    = 0
vertical binning      = 0
payload               = 2228224 bytes
exposure              = 0 ┬╡s
gain                  = 0 dB
[GvDevice::create_stream] Number of stream channels = 1
[GvDevice::auto_packet_size] Try packet size = 1500
[GvDevice::auto_packet_size] Try packet size = 773
[GvDevice::auto_packet_size] Try packet size = 409
[GvDevice::auto_packet_size] Try packet size = 227
[GvDevice::auto_packet_size] Try packet size = 136
[GvDevice::auto_packet_size] Try packet size = 91
[GvDevice::auto_packet_size] Try packet size = 68
[GvDevice::auto_packet_size] Try packet size = 57
[GvDevice::auto_packet_size] Packet size set to 1500 bytes
[GvStream::stream_new] Packet size = 1500 byte(s)
[GvStream::stream_new] Destination stream port = 45342
[GvStream::stream_new] Source stream port = 36659
[GvStream::loop] Standard socket method
gv n_stream channels  = 1
gv current channel    = 0
gv packet delay       = 0 ns
gv packet size        = 1500 bytes
[GvDevice::read_register] g_poll timed out after 500 ms
[GvDevice::read_register] Ack reception timeout
[GvDevice::read_register] g_poll timed out after 1 ms
[GvDevice::read_register] Ack reception timeout
[GvDevice::read_register] g_poll timed out after 0 ms
[GvDevice::read_register] Ack reception timeout
  0 frame/s  -       0 MiB/s - 2 errors
[GvDevice::read_register] g_poll timed out after 500 ms
[GvDevice::read_register] Ack reception timeout
[GvDevice::read_register] g_poll timed out after 1 ms
[GvDevice::read_register] Ack reception timeout
[GvDevice::read_register] g_poll timed out after 0 ms
[GvDevice::read_register] Ack reception timeout
[GvDevice::Heartbeat] Ack value = 2
  0 frame/s  -       0 MiB/s - 15 errors
  0 frame/s  -       0 MiB/s - 20 errors
[GvDevice::read_register] g_poll timed out after 500 ms
[GvDevice::read_register] Ack reception timeout
[GvDevice::read_register] g_poll timed out after 0 ms
[GvDevice::read_register] Ack reception timeout
[GvDevice::read_register] g_poll timed out after 500 ms
[GvDevice::read_register] Ack reception timeout
[GvDevice::read_register] g_poll timed out after 0 ms
[GvDevice::read_register] Ack reception timeout
  0 frame/s  -       0 MiB/s - 26 errors
[GvDevice::read_register] g_poll timed out after 500 ms
[GvDevice::read_register] Ack reception timeout
[GvDevice::read_register] g_poll timed out after 0 ms
[GvDevice::read_register] Ack reception timeout
[GvDevice::read_register] g_poll timed out after 500 ms
[GvDevice::read_register] Ack reception timeout
[GvDevice::read_register] g_poll timed out after 0 ms
[GvDevice::read_register] Ack reception timeout
  0 frame/s  -       0 MiB/s - 32 errors
[GvDevice::read_register] g_poll timed out after 500 ms
[GvDevice::read_register] Ack reception timeout
[GvDevice::read_register] g_poll timed out after 0 ms
[GvDevice::read_register] Ack reception timeout
[GvDevice::read_register] g_poll timed out after 500 ms
[GvDevice::read_register] Ack reception timeout
[GvDevice::read_register] g_poll timed out after 0 ms
[GvDevice::read_register] Ack reception timeout
  0 frame/s  -       0 MiB/s - 23 errors
[GvDevice::read_register] g_poll timed out after 500 ms
[GvDevice::read_register] Ack reception timeout
[GvDevice::read_register] g_poll timed out after 0 ms
[GvDevice::read_register] Ack reception timeout
[GvDevice::read_register] Unexpected answer (0x0000)
[GvDevice::read_register] g_poll timed out after 320 ms
[GvDevice::read_register] Ack reception timeout
[GvDevice::read_register] g_poll timed out after 0 ms
[GvDevice::read_register] Ack reception timeout
  0 frame/s  -       0 MiB/s - 16 errors
[GvDevice::read_register] g_poll timed out after 500 ms
[GvDevice::read_register] Ack reception timeout
[GvDevice::read_register] g_poll timed out after 0 ms
[GvDevice::read_register] Ack reception timeout
[GvDevice::read_register] g_poll timed out after 500 ms
[GvDevice::read_register] Ack reception timeout
[GvDevice::read_register] g_poll timed out after 0 ms
[GvDevice::read_register] Ack reception timeout
[GvDevice::Heartbeat] Ack value = 0
[GvDevice::Heartbeat] Tried 2 times
[GvDevice::Heartbeat] Control access lost
Control lost
  0 frame/s  -       0 MiB/s - 21 errors
Completed buffers = 0
Failures          = 155
Underruns         = 0
[GvDevice::write_register] g_poll timed out after 500 ms
[GvDevice::write_register] Ack reception timeout
[GvDevice::write_register] g_poll timed out after 0 ms
[GvDevice::write_register] Ack reception timeout
[GvDevice::write_register] g_poll timed out after 500 ms
[GvDevice::write_register] Ack reception timeout
[GvDevice::write_register] g_poll timed out after 0 ms
[GvDevice::write_register] Ack reception timeout
[GvDevice::write_register] Unexpected answer (0x0000)
[GvStream::loop] g_poll timed out after 20 ms
[GvStream::loop] g_poll timed out after 20 ms
[GvStream::loop] g_poll timed out after 20 ms
[GvStream::loop] g_poll timed out after 20 ms
[GvStream::loop] g_poll timed out after 20 ms
[GvDevice::write_register] g_poll timed out after 164 ms
[GvDevice::write_register] Ack reception timeout
[GvDevice::write_register] g_poll timed out after 0 ms
[GvDevice::write_register] Ack reception timeout
[GvDevice::write_register] g_poll timed out after 500 ms
[GvDevice::write_register] Ack reception timeout
[GvDevice::write_register] g_poll timed out after 0 ms
[GvDevice::write_register] Ack reception timeout
[GvDevice::write_register] Unexpected answer (0x0000)
[GvStream::loop] g_poll timed out after 1000 ms
[GvDevice::write_register] g_poll timed out after 297 ms
[GvDevice::write_register] Ack reception timeout
[GvDevice::write_register] g_poll timed out after 0 ms
[GvDevice::write_register] Ack reception timeout
  bins  ;Buffer r
       0;       0
    2000;       0
    4000;       0
    6000;       0
    8000;       0
   10000;       0
   12000;       0
   14000;       0
   16000;       0
   18000;       0
   20000;       0
   22000;       0
   24000;       0
   26000;       0
   28000;       0
   30000;       0
   32000;       0
   34000;       0
   36000;       0
   38000;       0
   40000;       0
   42000;       0
   44000;       0
   46000;       0
   48000;       0
   50000;       0
   52000;       0
   54000;       0
   56000;       0
   58000;       0
   60000;       0
   62000;       0
   64000;       0
   66000;       0
   68000;       0
   70000;       0
   72000;       0
   74000;       0
   76000;       0
   78000;       0
   80000;       0
   82000;       0
   84000;       0
   86000;       0
   88000;       0
   90000;       0
   92000;       0
   94000;       0
   96000;       0
   98000;       0
  100000;       0
  102000;       1
  104000;       1
  106000;       2
  108000;       0
  110000;       2
  112000;       0
  114000;       2
  116000;       0
  118000;       4
  120000;       1
  122000;       1
  124000;       1
  126000;       0
  128000;       1
  130000;       0
  132000;       3
  134000;       0
  136000;       3
  138000;       1
  140000;       0
  142000;       1
  144000;       1
  146000;       1
  148000;       0
  150000;       1
  152000;       0
  154000;       0
  156000;       1
  158000;       0
  160000;       2
  162000;       0
  164000;       1
  166000;       2
  168000;       1
  170000;       0
  172000;       0
  174000;       1
  176000;       1
  178000;       0
  180000;       0
  182000;       0
  184000;       2
  186000;       0
  188000;       1
  190000;       0
  192000;       1
  194000;       0
  196000;       0
  198000;       1
-------------
>=200000;     139
<      0;       0
min     ;  103881
max     ; 1260685
last max
at:     ;       0
Counter =        0
[GvStream::finalize] n_completed_buffers    = 0
[GvStream::finalize] n_failures             = 186
[GvStream::finalize] n_timeouts             = 186
[GvStream::finalize] n_aborteds             = 0
[GvStream::finalize] n_underruns            = 0
[GvStream::finalize] n_missing_frames       = 76
[GvStream::finalize] n_size_mismatch_errors = 0
[GvStream::finalize] n_received_packets     = 75880
[GvStream::finalize] n_missing_packets      = 281786
[GvStream::finalize] n_error_packets        = 0
[GvStream::finalize] n_ignored_packets      = 44265
[GvStream::finalize] n_resend_requests      = 2511901
[GvStream::finalize] n_resent_packets       = 2261
[GvStream::finalize] n_resend_ratio_reached = 105
[GvStream::finalize] n_duplicated_packets   = 1646
[Stream::finalize] Flush 50 buffer[s] in input queue
[Stream::finalize] Flush 0 buffer[s] in output queue
```

**Comment by eudoxos** [comment]
Juas as a note, I replace all calls to `g_get_monotonic_time` with thin wrapper calling `QueryPerformanceCounter` under windows (the frequency is 10M(Hz) in my case); the result is the same: no buffer completed, all timing out.
```c
gint64 arv_monotonic_time_us (void)
{
	#ifdef G_OS_WIN32
		static LARGE_INTEGER freq = { .QuadPart = 0 };
		LARGE_INTEGER t;

		if (freq.QuadPart == 0) { QueryPerformanceFrequency(&freq); }
		QueryPerformanceCounter(&t);
		return (t.QuadPart*1000000) / freq.QuadPart;
	#else
		return g_get_monotonic_time();
	#endif
}
```

**Comment by eudoxos** [comment]
I am happy to announce that this (as of the latest ea6d866) works very well on a native windows machine; using QueryPerformanceCounter (the snippet above) was not necessary.

There must be some other issues when running under Wine or in Virtualbox (networking? I have a fairly untypical LAN setup with secondary IP addresses, LLA and others; not sure how they deal with that). These issues can be looked at later, once this is merged.

Please merge this branch before it diverges from master.

I will write later some documentation on how to cross-compile aravis.

```
Z:\home\eudoxos\.local\share\crossroad\roads\w64\ara1\bin>arv-camera-test.exe
Looking for the first available camera
vendor name           = VRmagicImaging
model name            = VRmLS1-GEV#128#018-660#C
device id             = EFD0AFMH07
image width           = 2048
image height          = 1088
horizontal binning    = 0
vertical binning      = 4235304
payload               = 2228224 bytes
exposure              = 0 ┬╡s
gain                  = 0 dB
gv n_stream channels  = 1
gv current channel    = 0
gv packet delay       = 0 ns
gv packet size        = 1500 bytes
  9 frames/s -    20.1 MiB/s
 10 frames/s -    22.3 MiB/s
 10 frames/s -    22.3 MiB/s
 10 frames/s -    22.3 MiB/s
 10 frames/s -    22.3 MiB/s
  9 frames/s -    20.1 MiB/s
 11 frames/s -    24.5 MiB/s
 10 frames/s -    22.3 MiB/s
Completed buffers = 79
Failures          = 0
Underruns         = 0
```

**Comment by EmmanuelP** [comment]
Hi,

Great news. Thanks a lot.

I will review the patch next week.

**Comment by EmmanuelP** [comment]
Hi Václav,

I have created a new pull request with a cleaner history here: https://github.com/AravisProject/aravis/pull/477

The readme file is not in the pull request, as you said you will work on a documentation update.

I'm then closing this one. Thanks again.

**Comment by EmmanuelP** [comment]
Hi @eudoxos 

Did you have a chance to work on the windows build documentation ?

**Comment by eudoxos** [comment]
Thanks for the ping, I just tested msys2 packaging and created pull request: https://github.com/msys2/MINGW-packages/pull/9151 . I could not test with real hardware yet; the viewer works nicely (below) though arv-fake-camera does not start (I will check this out later).

The cross-build is documented in the [README-mingw.md](https://github.com/eudoxos/aravis/blob/619833ac37354a195d494ac5e58fe2f3f602d740/README-mingw.md) in this PR only, perhaps I should update it. And then, were would it go?

image

**Comment by eudoxos** [comment]
MSYS2 packages are already published: https://packages.msys2.org/base/mingw-w64-aravis :) . Would you like me to document the cross-building?

**Comment by EmmanuelP** [comment]
> MSYS2 packages are already published: https://packages.msys2.org/base/mingw-w64-aravis :) . Would you like me to document the cross-building?

I gave a try at building aravis using README-mingw.md, but it failed at the `crossroad install` step.

Perhaps an installation document about compiling aravis on a Windows machine would be simpler.

Or even better, a github action that would exercize the compilation on Windows. It would also help to catch possible regressions on Windows.

I tried the mingw packages. The performance was not really nice using the GigEVision protocol, there was a lots of missing packets. Using a USB3Vision device, after fiddling with zadig, I was able to see the camera, but streaming failed.

**Comment by EmmanuelP** [comment]
FWIW I have created a new issue in order to track the building documentation for Windows #533 

And also created a project about Windows support https://github.com/AravisProject/aravis/projects/3

**Comment by eudoxos** [comment]
> > MSYS2 packages are already published: https://packages.msys2.org/base/mingw-w64-aravis :) . Would you like me to document the cross-building?
> 
> I gave a try at building aravis using README-mingw.md, but it failed at the `crossroad install` step.

I think that might be related to an update of MSYS2 repos in crossroad, 2 weeks ago. You need to use upstream git version from https://git.tuxfamily.org/crossroad/crossroad.git/log/ ; pypi package is lagging and is thus dysfunctional.

**Pull request closed without merging**
